### DISCUSSION TITLE
2023-04: 为什么你该试试 Sccache？

### DISCUSSION THREAD

**Discussion by giscus** [Xuanwo/blog#395] [Announcements]
# 2023-04: 为什么你该试试 Sccache？

An infrastructure engineer, focused on distributed storage system

https://xuanwo.io/reports/2023-04/

**Comment by Endle** [comment]
使用了 sccache-action，github action里是不是就不用写丑陋的一大坨
```
      - name: Cache cargo registry
        uses: actions/cache@v2
        continue-on-error: false
        with:
          path: |
            ~/.cargo/registry
            ~/.cargo/git
          key: rust-${{ runner.os }}-cargo-${{ matrix.rust }}-${{ hashFiles('**/Cargo.toml') }}-build
          restore-keys: |
            rust-${{ runner.os }}-cargo-${{ matrix.rust }}-${{ hashFiles('**/Cargo.toml') }}-
            rust-${{ runner.os }}-cargo-${{ matrix.rust }}-
            rust-${{ runner.os }}-cargo-
            rust-
```

代码摘自 https://github.com/Endle/fireSeqSearch/blob/6e760731e1f91df5f647f0bd551aceb5d83bfcbb/.github/workflows/rust.yml#L30

**Reply by Xuanwo** [reply]
是的

**Reply by Xuanwo** [reply]
不过 sccache 并不会 cache cargo registry 相关的内容，可能需要对比一下两种方式的差异

**Reply by Endle** [reply]
要不我写一篇博文，讲一下我给我的个人项目换用 sccache 的经历，接着你在推特上转发一下？（图穷匕见x2) 我也安利一下我的fireSeqSearch

**Reply by Xuanwo** [reply]
好！

**Reply by Endle** [reply]
刚才搜了一下，找到了一篇2021年的博文 https://www.infinyon.com/blog/2021/04/github-actions-best-practices/

想麻烦一下 Xuanwo，你觉得这篇文章使用 sccache 的方式正确、或者说值得学习吗？或者这么问，使用你文中的 Action，是不是比按照这篇博文的内容更优雅一些？

**Reply by Xuanwo** [reply]
> 想麻烦一下 Xuanwo，你觉得这篇文章使用 sccache 的方式正确、或者说值得学习吗？或者这么问，使用你文中的 Action，是不是比按照这篇博文的内容更优雅一些？

文章里的方法本质上还是 rust-cache 那一套，每次都要全量上传全量加载

**Reply by Endle** [reply]
> 每次都要全量上传全量加载

你的意思是说，用你在文中提供的方法，可以不用上传整个 `~/.cache/sccache`, 而仅仅是这个文件夹的增量部分？

**Reply by Xuanwo** [reply]
> 你的意思是说，用你在文中提供的方法，可以不用上传整个 `~/.cache/sccache`, 而仅仅是这个文件夹的增量部分？

嗯，sccache 的工作原理是缓存编译产物，只有真的要用到才会上传/读取

**Comment by zu1k** [comment]
用起来啦 [book-searcher-org/book-searcher](https://github.com/book-searcher-org/book-searcher/blob/cc625c26383570b6b0c5e7778f2b4209cc6bfcfb/.github/workflows/test.yaml#L54), 效果很好

**Comment by yuanyin8811** [comment]
sccache config中size配置为20G, 存储介质为nas，编译时报
17:44:15  sccache: encountered fatal error
17:44:15  sccache: error: io error: unexpected end of file
17:44:15  sccache: caused by: io error: unexpected end of file 是什么原因呢
